File: Python/day06/main.py
import math
from functools import reduce


def load_data_part_1(filename: str) -> tuple[list[int], list[int]]:
    with open(filename) as f:
        times = list(map(int, f.readline().strip().split(":")[1].strip().split()))
        distances = list(map(int, f.readline().strip().split(":")[1].strip().split()))
        return times, distances


def load_data_part_2(filename: str) -> tuple[int, int]:
    with open(filename) as f:
        time = int(f.readline().split(":")[1].strip().replace(" ", ""))
        distance = int(f.readline().split(":")[1].strip().replace(" ", ""))
        return time, distance


def solution_part_1():
    times, distances = load_data_part_1("input.txt")
    ranges: list[int] = []
    for time, distance in zip(times, distances):
        ranges.append(calculate_range(distance, time))
    print(reduce(lambda x,y: x*y, ranges, 1))


def solution_part_2():
    time, distance = load_data_part_2("input.txt")
    print(calculate_range(distance, time))


def calculate_range(distance, time):
    p_2 = time / 2
    d = math.sqrt(math.pow(time / 2.0, 2.0) - distance)
    zero_1 = p_2 - d
    zero_2 = p_2 + d
    high = math.floor(zero_2)
    low = math.ceil(zero_1)
    if high == zero_2:
        high -= 1
    if low == zero_1:
        low += 1
    return high - low + 1


def main():
    solution_part_1()
    solution_part_2()

if __name__ == "__main__":
    main()

File: Rust/day06/src/main.rs
use std::fs;

fn load_data_part_1(filename: &str) -> (Vec<f64>, Vec<f64>) {
    let data = fs::read_to_string(filename).expect("Unable to read file");
    let times = data.lines().nth(0).unwrap()
        .split(":").nth(1).unwrap().trim().split_whitespace()
        .map(|x| x.parse::<f64>().unwrap()).collect::<Vec<f64>>();
    let distances = data.lines().nth(1).unwrap()
        .split(":").nth(1).unwrap().trim().split_whitespace()
        .map(|x| x.parse::<f64>().unwrap()).collect::<Vec<f64>>();
    return (times, distances)
}

fn load_data_part_2(filename: &str) -> (f64, f64) {
    let data = fs::read_to_string(filename).expect("Unable to read file");
    let time = data.lines().nth(0).unwrap()
        .split(":").nth(1).unwrap().trim().replace(" ", "").parse::<f64>().unwrap();
    let distance = data.lines().nth(1).unwrap()
        .split(":").nth(1).unwrap().trim().replace(" ", "").parse::<f64>().unwrap();
    return (time, distance)
}

fn solution_part_1() {
    let (times, distances) = load_data_part_1("input.txt");
    let mut ranges: Vec<f64> = Vec::new();
    for (time, distance) in times.iter().zip(distances.iter()) {
        let range = calculate_range(time, distance);
        ranges.push(range);
    }
    let product = ranges.into_iter().reduce(|x, y| x * y).unwrap();
    println!("Product: {}", product);
}

fn solution_part_2() {
    let (time, distance) = load_data_part_2("input.txt");
    let range = calculate_range(&time, &distance);
    println!("Range: {}", range);
}

fn calculate_range(time: &f64, distance: &f64) -> f64 {
    let zero_one = time / 2.0 + ((time / 2.0).powf(2.0) - distance).sqrt();
    let zero_two = time / 2.0 - ((time / 2.0).powf(2.0) - distance).sqrt();
    let mut high = zero_one.floor();
    if high == zero_one {
        high -= 1.0;
    }
    let mut low = zero_two.ceil();
    if low == zero_two {
        low += 1.0;
    }
    let range = high - low + 1.0;
    range
}

fn main() {
    solution_part_1();
    solution_part_2();
}
